step/step.go

```diff
@@ -0,0 +1,553 @@
+	HeadlessMode              bool `env:"headless_mode,opt[yes,no]"`
+
+	TestOptions         string `env:"xcodebuild_test_options"`
+	XcprettyTestOptions string `env:"xcpretty_test_options"`
+
+	// Debug
+	Verbose                     bool   `env:"verbose,opt[yes,no]"`
+	CollectSimulatorDiagnostics string `env:"collect_simulator_diagnostics,opt[always,on_failure,never]"`
+
+	// Output export
+	DeployDir             string `env:"BITRISE_DEPLOY_DIR"`
+	ExportUITestArtifacts bool   `env:"export_uitest_artifacts,opt[true,false]"`
+
+	CacheLevel string `env:"cache_level,opt[none,swift_packages]"`
+}
+
+type exportCondition int
+
+const (
+	invalid exportCondition = iota
+	always
+	never
+	onFailure
+)
+
+func parseExportCondition(condition string) exportCondition {
+	switch condition {
+	case "always":
+		return always
+	case "never":
+		return never
+	case "on_failure":
+		return onFailure
+	default:
+		return invalid
+	}
+}
+
+// Config ...
+type Config struct {
+	ProjectPath string
+	Scheme      string
+	TestPlan    string
+
+	XcodeMajorVersion int
+	SimulatorID       string
+	IsSimulatorBooted bool
+
+	TestRepetitionMode            string
+	MaximumTestRepetitions        int
+	RelaunchTestForEachRepetition bool
+
+	OutputTool         string
+	IsCleanBuild       bool
+	IsSingleBuild      bool
+	BuildBeforeTesting bool
+
+	RetryTestsOnFailure       bool
+	DisableIndexWhileBuilding bool
+	GenerateCodeCoverageFiles bool
+	HeadlessMode              bool
+
+	XcodebuildTestOptions string
+	XcprettyOptions       string
+
+	SimulatorDebug exportCondition
+
+	DeployDir             string
+	ExportUITestArtifacts bool
+
+	CacheLevel string
+}
+
+// XcodeTestRunner ...
+type XcodeTestRunner struct {
+	inputParser       stepconf.InputParser
+	logger            log.Logger
+	xcprettyInstaller xcpretty.Installer
+	xcodebuild        xcodebuild.Xcodebuild
+	simulator         simulator.Simulator
+	cache             cache.Cache
+	outputExporter    output.Exporter
+	pathModifier      pathutil.PathModifier
+}
+
+// NewXcodeTestRunner ...
+func NewXcodeTestRunner(inputParser stepconf.InputParser, logger log.Logger, xcprettyInstaller xcpretty.Installer, xcodebuild xcodebuild.Xcodebuild, simulator simulator.Simulator, cache cache.Cache, outputExporter output.Exporter, pathModifier pathutil.PathModifier) XcodeTestRunner {
+	return XcodeTestRunner{
+		inputParser:       inputParser,
+		logger:            logger,
+		xcprettyInstaller: xcprettyInstaller,
+		xcodebuild:        xcodebuild,
+		simulator:         simulator,
+		cache:             cache,
+		outputExporter:    outputExporter,
+		pathModifier:      pathModifier,
+	}
+}
+
+// ProcessConfig ...
+func (s XcodeTestRunner) ProcessConfig() (Config, error) {
+	var input Input
+	err := s.inputParser.Parse(&input)
+	if err != nil {
+		return Config{}, err
+	}
+
+	stepconf.Print(input)
+	s.logger.Println()
+
+	s.logger.EnableDebugLog(input.Verbose)
+
+	// validate Xcode version
```

Comment:
I'd extract validations to a separate method (or methods)

Reply:
separated `validateXcodeVersion` and `validateSimulator` in https://github.com/bitrise-steplib/steps-xcode-test/pull/181/commits/2941aefb06dc44f95e1ab40d083daf5467e89f63